app/controllers/spree/orders_controller_decorator.rb

```diff
@@ -139,17 +139,21 @@ def update_totals_and_taxes
   end
 
   # Sets the adjustments to open to perform the block's action and restores
-  # their state to whatever the they had.
+  # their state to whatever the they had. Note that it does not change any new
+  # adjustments that might get created in the yielded block.
   def with_open_adjustments
     previous_states = @order.adjustments.each_with_object({}) do |adjustment, hash|
-      hash[adjustment.id] = adjustment.state
+      hash[adjustment.id] = { adjustment: adjustment, previous_state: adjustment.state }
     end
     @order.adjustments.each(&:open)
 
     yield
 
-    @order.adjustments.each_with_index do |adjustment, index|
-      adjustment.update_attribute(:state, previous_states[adjustment.id])
+    previous_states.each do |adjustment_id, adjustment_pair|
+      adjustment = adjustment_pair[:adjustment]
+      previous_state = adjustment_pair[:previous_state]
+
+      adjustment.update_attribute(:state, previous_state)
```

Comment:
Now you have to hope that nothing deleted the adjustment or update_attribute will fail. :-) I thought of something like this:

```rb
  def with_open_adjustments
    previous_states = @order.adjustments.each_with_object({}) do |adjustment, hash|
      hash[adjustment.id] = adjustment.state
    end
    @order.adjustments.each(&:open)

    yield

    @order.adjustments.each do |adjustment|
      previous_state = previous_states[adjustment.id]
      adjustment.update_attribute(:state, previous_state) if previous_state
    end
end
```